### PR TITLE
GH-193 Adds support for dynamic base url

### LINKED ISSUE #193: Add support for dynamic base url
As a developer
I want to have support for dynamic base url
So that it is possible to change baseUrl after Rest instance is constructed

Main usecase scenario is to use dynamic config provider (aka config service) to get base url which may change over time/after config update.

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Fixes #193

### CODE REVIEW

**Comment by jhorstmann** [comment]
:+1:

File: riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java [hunk 1]
```diff
@@ -59,7 +60,11 @@ public RestBuilder baseUrl(@Nullable final String baseUrl) {
     }
 
     public RestBuilder baseUrl(@Nullable final URI baseUrl) {
-        this.baseUrl = baseUrl;
+        return baseUrlProvider(() -> baseUrl);
+    }
+
+    public RestBuilder baseUrlProvider(final Supplier<URI> baseUrlProvider) {
```
Comment (AlexanderYastrebov) [inline riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java]: This method could be named `baseUrl` as well

File: riptide-core/src/main/java/org/zalando/riptide/Rest.java [hunk 2]
```diff
@@ -5,24 +5,26 @@
...
     private final RequestArguments arguments;
     private final Plugin plugin;
 
     Rest(final AsyncClientHttpRequestFactory requestFactory, final List<HttpMessageConverter<?>> converters,
-            @Nullable final URI baseUrl, final Plugin plugin) {
+            final Supplier<URI> baseUrlProvider, final Plugin plugin) {
         this.requestFactory = checkNotNull(requestFactory, "request factory");
         this.worker = new MessageWorker(converters);
-        this.arguments = RequestArguments.create().withBaseUrl(baseUrl);
+        this.baseUrlProvider = checkNotNull(baseUrlProvider, "baseUrl provider");
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/Rest.java]: `"base url provider"`

File: riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java (hunk 1 above)
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java]: Yes, please.

File: riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import static org.zalando.riptide.Navigators.series;
+import static org.zalando.riptide.Route.pass;
+
+public class DynamicBaseUrlTest {
+
+    private final Rest unit;
+    private final MockRestServiceServer server;
+
+    @SuppressWarnings("unchecked")
+    private final Supplier<URI> baseUrlProviderMock = mock(Supplier.class);
```
Comment (whiskeysierra) [inline riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java]: I'd just use an `AtomicReference` as `baseUrl::get`. It's easier to use in your test.

File: riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void before() {
+        reset(baseUrlProviderMock);
+    }
+
+    @After
+    public void after() {
+        server.verify();
```
Comment (whiskeysierra) [inline riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java]: I'd move this into the test.

File: riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        final MockSetup setup = new MockSetup();
+
+        this.unit = setup.getRestBuilder().baseUrlProvider(baseUrlProviderMock).build();
+        this.server = setup.getServer();
+    }
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void before() {
+        reset(baseUrlProviderMock);
```
Comment (whiskeysierra) [inline riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java]: Not needed.

File: riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java (hunk 3 above)
Comment (AlexanderYastrebov) [inline riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java]: I want to verify number of calls to it as well

Comment (whiskeysierra) [inline riptide-core/src/test/java/org/zalando/riptide/DynamicBaseUrlTest.java]: Is that really important for your test?`If yes, I'd suggest to use a second test for that.

**Comment by whiskeysierra** [comment]
👍

**Pull request merged**